File: iza-os-intelligence/lib/router.ts
/**
 * LiteLLM to Ollama fallback router for IZA OS Intelligence Core
 */

export interface RouterConfig {
  liteLlmUrl: string;
  ollamaUrl: string;
  defaultModel: string;
  fallbackModel: string;
}

export class ModelRouter {
  private config: RouterConfig;

  constructor(config?: Partial<RouterConfig>) {
    this.config = {
      liteLlmUrl: process.env.LITELLM_URL || 'http://100.87.214.70:4000',
      ollamaUrl: process.env.OLLAMA_HOST || 'http://100.87.214.70:11434',
      defaultModel: 'qwen2.5:72b',
      fallbackModel: 'qwen2.5:7b',
      ...config,
    };
  }

  /**
   * Route request to LiteLLM with automatic local Ollama fallback on failure
   */
  async route(prompt: string, options: { model?: string } = {}): Promise<string> {
    const model = options.model || this.config.defaultModel;
    
    // 1. Try LiteLLM gateway first
    try {
      console.log(`[Router] Attempting LiteLLM gateway: ${this.config.liteLlmUrl} (Model: ${model})`);
      const response = await fetch(`${this.config.liteLlmUrl}/v1/chat/completions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: model,
          messages: [{ role: 'user', content: prompt }]
        })
      });
      
      if (response.ok) {
        const data = await response.json();
        return data.choices[0].message.content;
      }
      throw new Error(`LiteLLM returned status ${response.status}`);
    } catch (err) {
      console.warn(`[Router] LiteLLM gateway failed: ${err.message}. Falling back to Ollama.`);
      
      // 2. Fallback to local Ollama runtime (cost-free, offline-ready)
      return this.fallbackToOllama(prompt);
    }
  }

  private async fallbackToOllama(prompt: string): Promise<string> {
    try {
      console.log(`[Router] Attempting local Ollama: ${this.config.ollamaUrl} (Model: ${this.config.fallbackModel})`);
      const response = await fetch(`${this.config.ollamaUrl}/api/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: this.config.fallbackModel,
          prompt: prompt,
          stream: false
        })
      });

      if (response.ok) {
        const data = await response.json();
        return data.response;
      }
      throw new Error(`Ollama returned status ${response.status}`);
    } catch (err) {
      console.error(`[Router] Offline fallback failed: ${err.message}. Model route unreachable.`);
      throw new Error(`No available LLM backend: ${err.message}`);
    }
  }
}

File: iza-os-intelligence/lib/router.py
import os
import urllib.request
import json

class ModelRouter:
    def __init__(self, lite_llm_url=None, ollama_url=None, default_model=None, fallback_model=None):
        self.lite_llm_url = lite_llm_url or os.getenv("LITELLM_URL", "http://100.87.214.70:4000")
        self.ollama_url = ollama_url or os.getenv("OLLAMA_HOST", "http://100.87.214.70:11434")
        self.default_model = default_model or "qwen2.5:72b"
        self.fallback_model = fallback_model or "qwen2.5:7b"

    def route(self, prompt, model=None):
        target_model = model or self.default_model
        
        # 1. Try LiteLLM gateway
        try:
            print(f"[Router] Attempting LiteLLM gateway: {self.lite_llm_url} (Model: {target_model})")
            req = urllib.request.Request(
                f"{self.lite_llm_url}/v1/chat/completions",
                data=json.dumps({
                    "model": target_model,
                    "messages": [{"role": "user", "content": prompt}]
                }).encode('utf-8'),
                headers={'Content-Type': 'application/json'}
            )
            with urllib.request.urlopen(req, timeout=5) as response:
                if response.status == 200:
                    data = json.loads(response.read().decode('utf-8'))
                    return data['choices'][0]['message']['content']
        except Exception as e:
            print(f"[Router] LiteLLM gateway failed: {e}. Falling back to Ollama.")

        # 2. Try Local Ollama fallback
        return self.fallback_to_ollama(prompt)

    def fallback_to_ollama(self, prompt):
        try:
            print(f"[Router] Attempting local Ollama: {self.ollama_url} (Model: {self.fallback_model})")
            req = urllib.request.Request(
                f"{self.ollama_url}/api/generate",
                data=json.dumps({
                    "model": self.fallback_model,
                    "prompt": prompt,
                    "stream": False
                }).encode('utf-8'),
                headers={'Content-Type': 'application/json'}
            )
            with urllib.request.urlopen(req, timeout=5) as response:
                if response.status == 200:
                    data = json.loads(response.read().decode('utf-8'))
                    return data['response']
        except Exception as e:
            print(f"[Router] Offline fallback failed: {e}. Model route unreachable.")
            raise RuntimeError(f"No available LLM backend: {e}")
